src/core/server/saved_objects/migrations/core/document_migrator.test.ts

```diff
@@ -579,6 +584,71 @@ describe('DocumentMigrator', () => {
     }
   });
 
+  it('logs error during `applyMigrations` that are outside of transform functions', () => {
+    const log = mockLogger;
+    const migrator = new DocumentMigrator({
+      ...testOpts(),
+      typeRegistry: createRegistry({
+        name: 'dog',
+        migrations: {
+          '1.2.3': (doc) => doc,
+        },
+      }),
+      log,
+    });
+    const higherVersionDoc = {
+      id: 'smelly',
+      type: 'dog',
+      attributes: {},
+      migrationVersion: {
+        dog: '76.12.1',
+      },
+    };
+
+    try {
+      migrator.migrate(_.cloneDeep(higherVersionDoc));
+      expect('Did not throw').toEqual('But it should have!');
+    } catch (error) {
+      expect(error.message).toMatchInlineSnapshot(
+        `"Document \\"smelly\\" has property \\"dog\\" which belongs to a more recent version of Kibana [76.12.1]. The last known version is [1.2.3]"`
+      );
+      expect(mockLogger.error).toHaveBeenCalledTimes(1);
+      const loggedError = loggingSystemMock.collect(mockLoggerFactory).error[0][0];
+      expect(loggedError).toMatchInlineSnapshot(
+        `"Error applying migrations to doc dog:smelly: Error: Document \\"smelly\\" has property \\"dog\\" which belongs to a more recent version of Kibana [76.12.1]. The last known version is [1.2.3]"`
+      );
+    }
+  });
+
+  it('only logs error from transform function once', () => {
+    const log = mockLogger;
+    const migrator = new DocumentMigrator({
+      ...testOpts(),
+      typeRegistry: createRegistry({
+        name: 'dog',
+        migrations: {
+          '1.2.3': () => {
+            throw new Error('Dang diggity!');
+          },
+        },
+      }),
+      log,
+    });
+    const failedDoc = {
+      id: 'smelly',
+      type: 'dog',
+      attributes: {},
+      migrationVersion: {},
+    };
+    try {
+      migrator.migrate(_.cloneDeep(failedDoc));
+      expect('Did not throw').toEqual('But it should have!');
+    } catch (error) {
+      expect(mockLogger.warn).toHaveBeenCalledTimes(1);
+      expect(mockLogger.error).not.toHaveBeenCalled();
```

Comment:
Error from within the wrapped transform functions are currently logged as `warnings`

https://github.com/elastic/kibana/blob/079ef0620691179649f8787bf8e62c0d34e7da87/src/core/server/saved_objects/migrations/core/document_migrator.ts#L339-L341

As the error is actually causing the migration to fail and the server to not start, I wonder if we shouldn't upgrade that to the `error` level, wdyt?

Reply:
Upgrade to error makes sense 👍